src/analysis/convex/basic.lean

```diff
@@ -351,6 +352,141 @@ convex_halfspace_gt (is_linear_map.mk complex.add_im complex.smul_im) _
 lemma convex_halfspace_im_lge (r : ℝ) : convex {c : ℂ | r ≤ c.im} :=
 convex_halfspace_ge (is_linear_map.mk complex.add_im complex.smul_im) _
 
+/- Convex combinations in intervals -/
+
+/-
+If x is in an Ioo, it can be expressed as a convex combination of the endpoints.
+-/
+lemma convex.mem_Ioo {a b x : α} (h : a < b) :
+    x ∈ Ioo a b ↔ ∃ (x_a x_b : α), 0 < x_a ∧ 0 < x_b ∧ x_a + x_b = 1 ∧ x_a * a + x_b * b = x :=
+begin
+  split,
+  { rintros ⟨h_ax, h_bx⟩,
+    by_cases hab : ¬a < b,
+    { exfalso; exact hab h },
+    { have hab' : a ≠ b := ne_of_lt h,
+      refine ⟨(b-x) / (b-a), (x-a) / (b-a), _⟩,
+      refine ⟨div_pos (by linarith) (by linarith), div_pos (by linarith) (by linarith),_,_⟩,
+      { calc
+          (b - x) / (b - a) + (x - a) / (b - a)
+                = (b - x + (x - a)) / (b - a)     : by rw[←add_div]
+           ...  = (b - a) / (b-a)                 : by abel
+           ...  = 1                               : div_self (sub_ne_zero.mpr (ne.symm hab')) },
+
+      { calc
+          (b - x) / (b - a) * a + (x - a) / (b - a) * b
+                = ((b - x) * a) / (b-a) + ((x-a)*b)/(b-a)   : by simp only [mul_div_right_comm]
+          ...   = ((b-x) * a + (x-a)*b) / (b-a)             : by rw[←add_div]
+          ...   = (x * (b - a)) / (b-a)                     : by ring
+          ...   = x                                         : mul_div_cancel x (sub_ne_zero.mpr
+                                                                    (ne.symm hab')), } } },
+  { rw[mem_Ioo],
+    rintros ⟨xa, xb, ⟨hxa, hxb, hxaxb, h₂⟩⟩,
+    rw[←h₂],
+    split,
+    { calc
+      a   = 1 * a             : by rw[one_mul]
+      ... = (xa + xb) * a     : by rw[hxaxb]
+      ... = xa * a + xb * a   : by rw[add_mul]
+      ... < xa * a + xb * b   : by linarith[((mul_lt_mul_left hxb).mpr h)] },
+
+    { calc
+      b   = 1 * b             : by rw[one_mul]
+      ... = (xa + xb) * b     : by rw[hxaxb]
+      ... = xa * b + xb * b   : by rw[add_mul]
```

Comment:
rather than repeating this calc block so many times, I'd suggest proving a lemma that takes `b` and `hxaxb` as arguments. Then you can use `rw my_lem b hxaxh, nlinarith`.

Reply:
Indeed that cleaned things up quite a bit.